### PR TITLE
ADWINBoostingClassifier

### PR DESCRIPTION
**Pull request by danielnowakassis**
@MaxHalford  you have asked about using different drift detectors. This class depends on the error rate of ADWIN to reset the worst classifier in the Ensemble, like ADWINBaggingClassifier that is already in River .

### CODE REVIEW

File: river/ensemble/boosting.py [hunk 1]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
         return y_proba
 
 
+class ADWINBoostingClassifier(AdaBoostClassifier):
+    """ADWIN Boosting classifier.
+    ADWIN Boosting [^1] is the online boosting method of Oza and Russell [^2]
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:124]:
```suggestion
    """ADWIN boosting classifier.
    
    ADWIN Boosting [^1] is the online boosting method of Oza and Russell [^2]
```

File: river/ensemble/boosting.py [hunk 2]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+    with the addition of the `ADWIN` algorithm as a change detector. If concept
+    drift is detected, the worst member of the ensemble (based on the error
+    estimation by ADWIN) is replaced by a new (empty) classifier.
+    Parameters
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:128]:
```suggestion
    estimation by ADWIN) is replaced by a new (empty) classifier.
    
    Parameters
```

File: river/ensemble/boosting.py [hunk 3]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+    ----------
+    model
+        The classifier to boost.
+    n_models
+        The number of models in the ensemble.
+    seed
+        Random number generator seed for reproducibility.
+    Examples
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:136]:
```suggestion
        Random number generator seed for reproducibility.
        
    Examples
```

File: river/ensemble/boosting.py [hunk 4]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+    ...     model=(
+    ...         preprocessing.StandardScaler() |
+    ...         linear_model.LogisticRegression()
+    ...     ),
+    ...     n_models=3,
+    ...     seed=42
+    ... )
+    >>> metric = metrics.F1()
+    >>> evaluate.progressive_val_score(dataset, model, metric)
+    F1: 87.41%
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:155]:
```suggestion
    --------
    
    >>> from river import datasets
    >>> from river import ensemble
    >>> from river import evaluate
    >>> from river import linear_model
    >>> from river import metrics
    >>> from river import preprocessing
    
    >>> dataset = datasets.Phishing()
    >>> model = ensemble.ADWINBoostingClassifier(
    ...     model=(
    ...         preprocessing.StandardScaler() |
    ...         linear_model.LogisticRegression()
    ...     ),
    ...     n_models=3,
    ...     seed=42
    ... )
    >>> metric = metrics.F1()
    
    >>> evaluate.progressive_val_score(dataset, model, metric)
    F1: 87.41%
```

File: river/ensemble/boosting.py [hunk 5]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+    References
+    ----------
+    [^1]: Albert Bifet, Geoff Holmes, Bernhard Pfahringer, Richard Kirkby,
+    and Ricard Gavaldà. "New ensemble methods for evolving data streams."
+    In 15th ACM SIGKDD International Conference on Knowledge Discovery and
+    Data Mining, 2009.
+    [^2]: Oza, N., Russell, S. "Online bagging and boosting."
+    In: Artificial Intelligence and Statistics 2001, pp. 105–112.
+    Morgan Kaufmann, 2001.
+    """
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:166]:
```suggestion
    Morgan Kaufmann, 2001.
    
    """
```

File: river/ensemble/boosting.py [hunk 6]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+
+    def __init__(self, model: base.Classifier, n_models=10, seed: int = None):
+        super().__init__(model, n_models, seed)
+        self._drift_detectors = [copy.deepcopy(drift.ADWIN()) for _ in range(self.n_models)]
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:170]:
```suggestion
        self._drift_detectors = [drift.ADWIN() for _ in range(self.n_models)]
```

No need to call `copy.clone` if you instantiate the class

File: river/ensemble/boosting.py [hunk 7]
```diff
@@ -118,6 +119,98 @@ def predict_proba_one(self, x):
...
+
+            try:
+                y_pred = model.predict_one(x)
+                error_estimation = self._drift_detectors[i].estimation
+                self._drift_detectors[i].update(int(y_pred == y))
+                if self._drift_detectors[i].drift_detected:
+                    if self._drift_detectors[i].estimation > error_estimation:
+                        change_detected = True
+            except ValueError:
+                change_detected = False
```
Comment (MaxHalford) [inline river/ensemble/boosting.py:200]: Can you comment where the `ValueError` is raised? I always find it a bit hacky to rely on an exception

**Review by MaxHalford — CHANGES_REQUESTED** [review]

**Comment by MaxHalford** [comment]
Thanks for answer @danielnowakassis! Looks good :)

Can you add a note to the `unreleased.md` file?

File: docs/releases/unreleased.md [hunk 8]
```diff
@@ -21,6 +21,8 @@
 
 ## ensemble
 
+- Added `ensemble.ADWINBoostingClassifier`.
+
```
Comment (MaxHalford) [inline docs/releases/unreleased.md:25]:
```suggestion
```

**Review by MaxHalford — CHANGES_REQUESTED** [review]

File: river/ensemble/boosting.py (hunk 7 above)
Comment (danielnowakassis) [reply river/ensemble/boosting.py:207]: This part of the code was based in the ADWINBagging code. I didn't find out what raise the ValueError...

**Pull request merged**